packages/react-urql/src/hooks/useQuery.ts

```diff
@@ -355,25 +357,30 @@ export function useQuery<
   }, [cache, state[0], state[2][1]]);
 
   const executeQuery = React.useCallback(
-    (opts?: Partial<OperationContext>) => {
-      const context = {
-        requestPolicy: args.requestPolicy,
-        ...args.context,
-        ...opts,
-      };
+    (opts?: Partial<OperationContext>) =>
+      new Promise<void>(resolve => {
+        const context = {
+          requestPolicy: args.requestPolicy,
+          ...args.context,
+          ...opts,
+        };
 
-      deferDispatch(setState, state => {
-        const source = suspense
-          ? pipe(
-              client.executeQuery(request, context),
-              onPush(result => {
-                cache.set(request.key, result);
-              })
-            )
-          : client.executeQuery(request, context);
-        return [source, state[1], deps];
-      });
-    },
+        deferDispatch(setState, state => {
+          const source = suspense
+            ? pipe(
+                client.executeQuery(request, context),
+                onPush(result => {
+                  cache.set(request.key, result);
+                  resolve();
+                })
+              )
+            : pipe(client.executeQuery(request, context), result => {
+                resolve();
```

Comment:
in a `cache-and-network` call this would most likely return the stale equivalent. We do already have a streaming util to convert this to a Promise conditionally in [https://github.com/urql-graphql/urql/blob/main/packages/core/src/utils/streamUtils.ts#L15].

So what we could do is create the `source`, do a `deferDispatch` to set the state to React and return the `source`